[MM-44315] Downgrade - Team selection flow

nickmisasi opened the pull request:
#### Summary
This PR adds support for downgrading to starter through the pricing modal. When the downgrade button is clicked, there are a few cases:
1. only one active team - they'll be downgraded
2. More than one active team - this PR's new modal will open
2.1. If there are 4 or more teams active, the modal will contain a drop down to select which team to keep
2.2. If there are fewer than 4 teams active, you will see radio buttons

Upon choosing a team (with either radio or drop down), clicking "downgrade", all teams but the one you selected will be archived. You will then be redirected to a success screen. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-43088
https://mattermost.atlassian.net/browse/MM-44315

#### Related Pull Requests
N/A

#### Screenshots
Trying to downgrade with 4+ active teams:
![image](https://user-images.githubusercontent.com/12176405/173865500-040ce451-e53b-456e-8be9-3a431af4cdaa.png)
Trying to downgrade with fewer than 4 teams:
![image](https://user-images.githubusercontent.com/12176405/173865712-97db469b-7a1c-4a98-bf63-a2ede04898b5.png)

Success screen: 
![image](https://user-images.githubusercontent.com/12176405/172907079-0e1372c2-e047-46ee-98db-226c1784d272.png)
Error screen:
![image](https://user-images.githubusercontent.com/12176405/173864307-da3a36db-9be0-4b23-a7de-33c3f3da76ae.png)


Button in the billing_subscriptions page to open plans modal
![image](https://user-images.githubusercontent.com/12176405/173623163-21c818d3-7bbd-4118-ab8e-a2f3d9a9b463.png)



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
None
```

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

nickmisasi commented:
@annelieseklein Some notes about this one:

I couldn't include the team name in the `Return to workspace` button on the success screen  as it is in the designs - so John and I decided to just go with `Return to workspace`

I made an executive decision on leaving the pricing modal open, and popping this one over top. If you'd like that behaviour to change, let me know.

neallred commented on components/success_modal/index.tsx:
```diff
@@ -0,0 +1,89 @@
+                formattedSubtitle={
+                    <FormattedMessage
+                        id={'success_modal.subtitle'}
+                        defaultMessage={'default message'}
```
Nit: replace with the actual default message. `"Your final bill will be prorated. Your workspace now has {planName} limits."`

If we don't pass the real default message, then for partially translated languages that haven't yet translated, they will see "default message" instead of the English variant.

neallred commented on i18n/en.json:
```diff
@@ -4517,6 +4524,8 @@
   "status_dropdown.set_ooo.extra": "Automatic Replies are enabled",
   "string.id": "default message",
+  "success_modal.return_to_workspace": "Return to workspace",
+  "success_modal.subtitle": "Your final bill will be prorated. Your workspace now has Cloud Starter limits.",
```
Can we make this `has Cloud Starter limits` -> `has {planName} limits`? It will be less translation churn when we also support downgrading from enterprise to professional.

neallred commented on components/pricing_modal/content.tsx:
```diff
@@ -241,9 +265,31 @@ function Content(props: ContentProps) {
+                                            },
+                                        }),
+                                    );
+                                } else {
```
Looks like the designs also call for a third state, where we open a variation of the limits modal if there is only one team, but one or more of the limits is exceeded: https://www.figma.com/file/Aob5ffHtfjKzQoebR93ycA/Cloud-Freemium?node-id=657%3A436901

![one-team-and-exceeded-limit](https://user-images.githubusercontent.com/13738432/172929405-b0b58f9a-807a-4f8d-80e3-99d170708932.jpg)

It would need to make use of the hardcoded `fallbackStarterLimits` and `fallbackProfessionalLimits` until https://mattermost.atlassian.net/browse/MM-44787 is implemented.

nickmisasi replied:
The ticket only covers the team downgrade: https://mattermost.atlassian.net/browse/MM-44315

I think because of what you mentioned, but also, we're not actually enforcing the archiving of anything else for the time being (AFAIK)

neallred replied:
sounds good to me!

nickmisasi commented:
@neallred 

>implementing the 1 team & exceeded limit case (or agreement to defer implementing)

I think https://github.com/mattermost/mattermost-webapp/pull/10452#discussion_r893896728 covers this?


>changing the message wording to include the message history limit and board card limit, and probably not show the board view limit

The current copy matches that of the design - @johndavidlugtu based on Nathaniel's suggestion, what do you think about:
```
Cloud starter is restricted to 1 team, 10,000 messages, 10GB file storage, 10 apps, and 500 board cards. If you downgrade, some data will be archived. It won't be deleted and you'll be able to access it again when you upgrade.
```

neallred commented:
> I think #10452 (comment) covers this?

:+1:

johndavidlugtu commented:
@nickmisasi @neallred good idea on mentioning message history and board limit. Just slightly changed the copy: 

Cloud starter is restricted to 1 team, 10,000 messages, 10GB file storage, 5 apps, and 500 board cards. If you downgrade, some data will be archived. Archived data can be accessible when you upgrade back.

It should say 5 apps and not 10 by the way

stevemudie commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/205625

annelieseklein commented:
> @annelieseklein Some notes about this one:
> 
> I couldn't include the team name in the `Return to workspace` button on the success screen as it is in the designs - so John and I decided to just go with `Return to workspace`
> 
> I made an executive decision on leaving the pricing modal open, and popping this one over top. If you'd like that behaviour to change, let me know.

@nickmisasi Yes, great, this one should pop over top but there should be an overlay in between as shown in the prototype

stevemudie commented:
@nickmisasi This _works_, but there a few things going on that you might want to look at.

- If I don't upgrade to Professional I can't add any teams as that's a Professional & Enterprise feature only.
- If I upgrade to Professional there isn't any way to ever see the pricing modal as all roads lead to the Contact Us path for upgrading to Enterprise. (The ticket linked above describes "[Downgrade from Professional](https://mattermost.atlassian.net/browse/MM-44315) which led me down that path to begin with.)"
**cc**: @annelieseklein @marianunez With this in mind I'm thinking the downgrade button is going to have to live somewhere other than the pricing modal? Perhaps on the `System Console ➜ Billing & Account ➜ Subscription` page? 🤔 

- If I go after the Enterprise trial it says "failed" but ultimately comes around and I do indeed have the Enterprise trial running.
- Once the Enterprise trial is in place then I still have access to the pricing modal but it will only let me add a single team.
- If I try to add a second team then things sort of fall apart but I'll spare you the details right now as I don't know if it's so much a bug as it is the nature of this particular Spinwick ATM.
- If I downgrade though it does play nice and it lets me pick the team (of the 2 I have) to keep and it saves just that team so that part works. 🎉 
- The actual downgrade happens lightning fast! But I did notice something weird on the page saying you were downgraded. I need to revisit that and I'll add more once I catch it again.

That's it in a nutshell and I'll get back on this in the morning. 🙃

nickmisasi commented:
Thanks for the review @stevemudie! Lot's to unpack here - will comment inline.

>If I don't upgrade to Professional then I can't add any teams as that's a Professional & Enterprise feature only.

This functionality is only present for downgrades, so this is exactly as planned. You'd have upgraded to Professional or Enterprise at some point in order to downgrade to Starter.

>If I upgrade to Professional there isn't any way to ever see the pricing modal as all roads lead to the Contact Us path for upgrading to Enterprise. (The ticket linked above describes "[Downgrade from Professional](https://mattermost.atlassian.net/browse/MM-44315) which led me down that path to begin with.)"
cc: @annelieseklein @marianunez With this in mind I'm thinking the downgrade button is going to have to live somewhere other than the pricing modal? Perhaps on the System Console ➜ Billing & Account ➜ Subscription page? 🤔

Mentioned this to Maria myself as well. This isn't going to be a June 15th feature, so we should be able to tackle this in the next release. Will defer to Maria/Anneliese on how that fleshes out

>If I go after the Enterprise trial it says "failed" but ultimately comes around and I do indeed have the Enterprise trial running.

This sounds like a bug unrelated to these changes. Most of what's here is net new code, and none of it surrounds the trial lifecycle. I was able to reproduce this once in a normal installation, so will look into this separately.

>Once the Enterprise trial is in place then I still have access to the pricing modal but it will only let me add a single team.

This is also a bug, which again I think is unrelated. I was able to reproduce in a test portal instance - refreshing fixes it, so likely the limits aren't being refreshed properly along with the subscription.

>If I downgrade though it does play nice and it lets me pick the team (of the 2 I have) to keep and it saves just that team so that part works. 🎉

This functionality was the focus of this PR, so I'm glad that it works!

For the other 2 I didn't address, let me know when you've got more details!

neallred reviewed (approved):
LGTM, thanks @nickmisasi !

AGMETEOR commented on components/dropdown_input.tsx:
```diff
@@ -9,7 +9,7 @@ import './dropdown_input.scss';
 
 // TODO: This component needs work, should not be used outside of AddressInfo until this comment is removed.
```
👀

nickmisasi replied:
Didn't even see that 😂 . Asked Devin (who wrote the comment) if he has any concerns

AGMETEOR commented on components/pricing_modal/downgrade_team_removal_modal/index.tsx:
```diff
@@ -0,0 +1,220 @@
+import './downgrade_team_removal_modal.scss';
+
+type Props = {
+    onHide?: () => void;
```
When opening this, only product_id and starterProductName are passed as dialogProps.

nickmisasi replied:
Removed

neallred commented on components/pricing_modal/content.tsx:
```diff
@@ -180,6 +188,22 @@ function Content(props: ContentProps) {
+        }
+
+        dispatch(subscribeCloudSubscription(starterProduct?.id));
+        dispatch(closeModal(ModalIdentifiers.CLOUD_DOWNGRADE_CHOOSE_TEAM));
```
is it still WIP or maybe didn't get pushed? I don't see it in the diff.

nickmisasi replied:
Sorry, a lot going on today haha I even said in planning I was going to do this but forgot. Will address asap

nickmisasi replied:
Addressed in latest commit

neallred commented on components/pricing_modal/downgrade_team_removal_modal/downgrade_team_removal_modal.test.tsx:
```diff
@@ -0,0 +1,357 @@
+                />
+            </redux.Provider>,
+        );
+        expect(screen.getByText('Confirm Plan Downgrade')).toBeInTheDocument();
```
This looks more "jesty", but 0/5 on this style vs without the expect. `screen.getByText` throws an error/fails the test if it does not exist: https://testing-library.com/docs/queries/about#types-of-queries . The more common case of needing to combine with a jest expect is when you are doing a `queryBy` for something that does not exist. Then you expect it `not.toBeInTheDocument()`

neallred commented on components/pricing_modal/content.tsx:
```diff
@@ -181,6 +190,34 @@ function Content(props: ContentProps) {
+
+        const result = await dispatch(subscribeCloudSubscription(starterProduct?.id));
+
+        if (typeof result === 'boolean') {
```
Nit: Even though this action currently only returns a true as a boolean, could we do a `if (typeof result === 'boolean') && result) {` here to future proof against if an additional error state that returns `false` is added?

neallred reviewed (approved):
LGTM! A couple non-blocking nits

johndavidlugtu commented:
@nickmisasi copy looks good re: error screen.

One thing observed though: "Cloud Free". Can we switch all mentions of "Cloud Free" into "Cloud Starter"?

johndavidlugtu reviewed (approved):
Looks good, thanks!

stevemudie commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/208464

AGMETEOR commented:
@nickmisasi looks good. Just need to update the current E2E to include this flow

AGMETEOR reviewed (commented):
The current E2E needs to include this flow

stevemudie reviewed (approved):
This is all set @nickmisasi though I did notice an odd error wording when I tried to create an existing team.

![Existing team](https://user-images.githubusercontent.com/7083912/174645195-e3186870-df12-4caa-9a3d-8b8f1c1663a7.png)

Thoughts?

**ETA**: A team I knew existed but had not yet been unarchived.

nickmisasi commented:
@stevemudie I can repro this locally on an on-prem instance - think the behaviour is intentional, to trigger the user to realize there's already a team with that name. Probably could use better copy

stevemudie commented:
> I can repro this locally on an on-prem instance - think the behaviour is intentional, to trigger the user to realize there's already a team with that name. Probably could use better copy

@nickmisasi Thanks. I wrote [a ticket](https://mattermost.atlassian.net/browse/MM-45237) to address it at some point.

annelieseklein commented:
Looks great! 

1. The "Confirm team" modal should have rounded corners that are all the same (The bottom corners have the right amount of roundedness)
design: <img width="459" alt="Screen Shot 2022-06-21 at 8 13 26 PM" src="https://user-images.githubusercontent.com/29821919/174937050-f02e0843-47e3-409a-a94a-b63cad411a27.png">
current implementation: <img width="350" alt="Screen Shot 2022-06-21 at 8 14 20 PM" src="https://user-images.githubusercontent.com/29821919/174937057-248ff40f-1f67-4c48-ae13-19d83ed8140f.png">

2. Can we have the "current plan" be 8 px and centered above the professional plan and have the CTA be disabled? see image 
<img width="589" alt="Screen Shot 2022-06-21 at 8 28 11 PM" src="https://user-images.githubusercontent.com/29821919/174937300-05dc3162-3ee2-4e4d-84b3-3710b6ff3e43.png">

3. Can we please center the icon with the text. The icon should be 14px, and the text is 11px semibold
design: 
<img width="398" alt="Screen Shot 2022-06-21 at 8 30 17 PM" src="https://user-images.githubusercontent.com/29821919/174937577-939ff021-7b80-4f79-9c3a-7b36986acd50.png">

Current implementation: <img width="291" alt="Screen Shot 2022-06-21 at 8 29 59 PM" src="https://user-images.githubusercontent.com/29821919/174937539-70e9fb67-f117-4d02-8d75-08abbccaeee9.png">

AGMETEOR reviewed (approved):
Pending uncommenting some code in the spec file. Otw, LGTM

nickmisasi commented:
@annelieseklein upper corners of the modal have been rounded, and I adjusted the error text (centered vertically, and adjusted both the icon and the text to better match the sizing in the design)
![image](https://user-images.githubusercontent.com/12176405/175053659-9fffeb8f-8d55-48b0-afaf-711f217e9c6b.png)

The "Current Plan" label appears to only be set up to show if the starter plan is the current plan. Think this was some missed logic in initial implementation. It's out of scope of this PR, so I've created a ticket for it: https://mattermost.atlassian.net/browse/MM-45254

annelieseklein commented:
> @annelieseklein upper corners of the modal have been rounded, and I adjusted the error text (centered vertically, and adjusted both the icon and the text to better match the sizing in the design) ![image](https://user-images.githubusercontent.com/12176405/175053659-9fffeb8f-8d55-48b0-afaf-711f217e9c6b.png)
> 
> The "Current Plan" label appears to only be set up to show if the starter plan is the current plan. Think this was some missed logic in initial implementation. It's out of scope of this PR, so I've created a ticket for it: https://mattermost.atlassian.net/browse/MM-45254

Awesome, thanks! Last thing, but the "Select a team" should be 16px away from the left border of the input field

annelieseklein reviewed (approved):
Looks awesome, thank you Nick!

AGMETEOR commented on components/cloud_subscribe_result_modal/style.scss:
```diff
@@ -0,0 +1,8 @@
+.IconMessage {
```
@nickmisasi 
<img width="1792" alt="Screenshot 2022-07-05 at 12 16 19" src="https://user-images.githubusercontent.com/28563179/177294737-ce92bfff-ddde-4fe3-aae1-77b432dcab7e.png">

Could be breaking because of this? 🤔

mickmister commented on components/cloud_subscribe_result_modal/style.scss:
```diff
@@ -0,0 +1,8 @@
+    transform: translateY(50%);
+}
+
+.FullScreenModal {
```
@nickmisasi @AGMETEOR This line is causing issues with opening a `ConfirmModal` on top of a `FullScreenModal`:

```js
.FullScreenModal {
    z-index: 99999 !important;
}
```

The plugin marketplace is implemented as a fullscreen modal, so the confirm modal to upgrade a plugin isn't showing https://mattermost.atlassian.net/browse/MM-45731

Any ideas on what next steps are to fix this?

mickmister replied:
Nevermind it's solved in https://github.com/mattermost/mattermost-webapp/pull/10702

Thanks @AGMETEOR

mickmister replied:
@nickmisasi The issue is that the change is currently in the released 7.1 version for self-managed. We'll need to include Allan's fix in a dot release

nickmisasi replied:
@mickmister This `z-index` was added in a cloud specific flow. Allan's removal 8 days ago  would have been tested as part of release testing for the cloud release today - if there were any significant issues resulting from the removal of the z-index, we'd have noticed today.
